packages/remirror__extension-link/src/link-extension-utils.ts

```diff
@@ -53,3 +53,59 @@ export const TOP_50_TLDS = [
   'no',
   'cyou',
 ];
+
+/** Including single and double quotation */
+export const SENTENCE_PUNCTUATIONS = ['!', '?', "'", ',', '.', ':', ';', '"'];
+
+export const DEFAULT_ADJACENT_PUNCTUATIONS = [...SENTENCE_PUNCTUATIONS, '(', ')', '[', ']'];
+
+const PAIR_PUNCTUATIONS = '[]{}()';
+
+export const isBalanced = (input: string): boolean => {
+  const stack = [];
+
+  for (const character of input) {
+    const index = PAIR_PUNCTUATIONS.indexOf(character);
+
+    if (index === -1) {
+      continue;
+    }
+
+    if (index % 2 === 0) {
+      stack.push(index + 1);
+
+      continue;
+    }
+
+    if (stack.length === 0 || stack.pop() !== index) {
+      return false;
+    }
+  }
+
+  return stack.length === 0;
+};
+
+export const getBalancedIndex = (input: string, index: number): number =>
+  !isBalanced(input.slice(0, index)) ? getBalancedIndex(input, --index) : index;
```

Comment:
I suspect this logic could be simplified by reversing it - instead of recursively calling `isBalanced`, work out where the imbalance occurs on the original string

FYI, this returns a positive index - from the beginning of the string - rather than a negative one from the end

```ts
export const getBalancedIndex = (str: string): number => {
  const stack: Array<{ index: number, expected: string }> = [];

  for (let i = 0; i < str.length; i++) {
    const char = str.charAt(i);
    const bracketIndex = PAIR_PUNCTUATIONS.indexOf(char);

    // char is not a bracket
    if (bracketIndex === -1) {
      continue;
    }


    // char is an open bracket character
    if (bracketIndex % 2 === 0) {
      stack.push({ index: i, expected: PAIR_PUNCTUATIONS.charAt(bracketIndex + 1) });
      continue;
    }

    // char is a closing bracket character
    const unmatched = stack.pop();

    if (unmatched === undefined) {
      // closing bracket without any opening bracket
      return i;
    }

    const { index, expected } = unmatched;

    if (char !== expected) {
      // closing bracket was not what we expected
      return index;
    }
  }

  const unmatched = stack.pop();

  if (unmatched === undefined) {
    // str is balanced
    return -1;
  }

  return unmatched.index;
}

export const isBalanced = (str: string): boolean => {
  return getBalancedIndex(str) === -1;
};
```

Reply:
Don't see the benefit of getting a positive index. Also, I am not sure if this will return the expected index without additional changes. I will test it though.

Reply:
If this is a blocker for you I am happy to investigate further but as of right now it is not always returning the correct index.

Reply:
@whawker I ended up implementing your suggestion after fixing the issue I noticed with it.

I was able to remove `isBalanced` and reduce the complexity in `getLinkEndIndex`.